cpu/samd5x/periph/pm.c

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2017 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_saml21
+ * @ingroup     drivers_periph_pm
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the kernels power management interface
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include "periph/pm.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+void pm_set(unsigned mode)
+{
+    if (mode < PM_NUM_MODES) {
+        uint32_t _mode;
+
+        switch (mode) {
+            case 0:
+                DEBUG("pm_set(): setting STANDBY mode.\n");
+                _mode = PM_SLEEPCFG_SLEEPMODE_STANDBY;
+                break;
+            default: /* Falls through */
+            case 1:
+                DEBUG("pm_set(): setting IDLE2 mode.\n");
+                _mode = PM_SLEEPCFG_SLEEPMODE_IDLE2;
+                break;
+            case 2:
+                DEBUG("pm_set(): setting IDLE1 mode.\n");
+                _mode = PM_SLEEPCFG_SLEEPMODE_IDLE1;
+                break;
+            case 3:
+                DEBUG("pm_set(): setting IDLE0 mode.\n");
+                _mode = PM_SLEEPCFG_SLEEPMODE_IDLE0;
+                break;
+        }
+
+        /* write sleep configuration */
+        PM->SLEEPCFG.bit.SLEEPMODE = _mode;
+        /* make sure value has been set */
+        while (PM->SLEEPCFG.bit.SLEEPMODE != _mode) {}
+    }
+
+    cortexm_sleep(mode);
```

Comment:
There are now four variants of this, all have a different approach to call `cortexm_sleep`. In this implementation it is important to identify that only `mode == 0` does a light sleep. The other modes do a deep sleep. I don't see it being documented.

Also the mapping 0->STANDBY, 1->IDLE2, 2->IDLE1, 3->IDLE0 needs explanation or documentation.

Reply:
Actually all modes but 0 do a light sleep - but just looking at the power consumption, `PM_SLEEPCFG_SLEEPMODE_IDLE2` is the default anyway if only `cortexm_sleep()` is called.

This MCU also supports three deep sleep states (STANDBY, HIBERNATE & BACKUP) but since pm_layered only supports 4 sleep modes (and on all other sam0 boards modes 1-3 are used for light sleep / idle and I wanted to keep the semantics the same), only the deepest one is exposed. The RTC still can wake the MCU in this mode which is good enough for me.

Since all the sam0s expose different sets of sleep states I think the pain of light code duplication is still less than that that of a confusing mess of `#ifdefs` in a common version.

Reply:
Ok, so BACKUP only works if Vbat is connected (but allows to retain some RAM) whereas HIBERNATE  always works, so I'm switching to that by default.